application/cache/cache.go

```diff
@@ -0,0 +1,233 @@
+package cache
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+
+	"github.com/go-redis/cache/v8"
+	"github.com/go-redis/redis/v8"
+
+	"github.com/silinternational/wecarry-api/api"
+	"github.com/silinternational/wecarry-api/conversions"
+	"github.com/silinternational/wecarry-api/models"
+)
+
+var (
+	RequestsCache         *cache.Cache
+	PublicRequestsKeyName string = "public"
+)
+
+func init() {
+	ring := redis.NewRing(&redis.RingOptions{
+		Addrs: map[string]string{
+			"redis": "redis:6379",
+		},
+	})
+
+	RequestsCache = cache.New(&cache.Options{
+		Redis:      ring,
+		LocalCache: cache.NewTinyLFU(1000, time.Minute),
+		// use json in lieu of msgpack for encoding/decoding
+		Marshal:   json.Marshal,
+		Unmarshal: json.Unmarshal,
+	})
+}
+
+// Wrapper for cache write
+func CacheWrite(ctx context.Context, organization string, requestsMap interface{}) error {
+	return RequestsCache.Set(&cache.Item{
+		Ctx:   ctx,
+		Key:   organization,
+		Value: requestsMap,
+		TTL:   time.Hour,
+	})
+}
+
+// Get all requests visible to user in specified organization from cache, fetching data from database if needed
+func GetVisibleRequests(ctx context.Context, orgaziation models.Organization) ([]api.RequestAbridged, error) {
+	// get private requests
+	_, privateRequestsMap, err := getOrCreateCacheEntryPrivate(ctx, orgaziation)
+	if err != nil {
+		return nil, err
+	}
+
+	// get public requests
+	_, publicRequestsMap, err := getOrCreateCacheEntryPublic(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// compose them to get list of all requests visible to the current user
+	visibleRequestsList := make([]api.RequestAbridged, len(publicRequestsMap)+len(privateRequestsMap))
+	for _, privateRequest := range privateRequestsMap {
+		visibleRequestsList = append(visibleRequestsList, privateRequest)
+	}
+	for _, publicRequest := range publicRequestsMap {
+		visibleRequestsList = append(visibleRequestsList, publicRequest)
+	}
+
+	return visibleRequestsList, nil
+}
+
+// Wrapper for cache read
+func CacheRead(ctx context.Context, organization string, requestsMap interface{}) error {
+	return RequestsCache.Get(ctx, organization, requestsMap)
+}
+
+// Rebuild cache after a request is created or updated
+func CacheRebuildOnChangedRequest(ctx context.Context, organization models.Organization, request models.Request) error {
+	// update private cache
+	newEntryCreated, requestsMap, err := getOrCreateCacheEntryPrivate(ctx, organization)
+	if err != nil {
+		return err
```

Comment:
It is generally preferred to add to the error so that debugging is less painful. Sometimes it adds no value, but usually it does. (e.g. `return errors.New("error in cache rebuild: "+err.Error())`)